[AMDGPU] Correct bitshift legality transformation for small vectors

zGoldthorpe opened the pull request:
Fix for a bug found by the AMD fuzzing project.

The legaliser would originally try to widen a small vector such as `<4 x i1>` to a single `i16` during the legalisation of bitshifts, as it was not originally written with consideration for vector operands. This patch simply adds a guard to prohibit this transformation and allow other legalisation transformations to step in.

zGoldthorpe committed:
Correct bitshift legalisation

zGoldthorpe commented:
@ronlieb requesting review

shiltian commented on llvm/test/CodeGen/AMDGPU/widen-vector-shift.ll:
```diff
@@ -0,0 +1,24 @@
+; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx90a -O0 -print-after=legalizer %s -o /dev/null 2>&1 | FileCheck %s
```
Use full triple `amdgcn-amd-amdhsa`. Auto generate check lines pls. Also use `-stop-after`.

shiltian replied:
Is it necessary to use `-O0` here?

arsenm replied:
checking stderr should be a last resort, but if you're discarding stdout you should use -filetype=null

arsenm replied:
Test also belongs in GlobalISel subdirectory, but preferably would just be adding new type cases to the existing instruction tests. You should also just run it to the end and check the isa output

zGoldthorpe replied:
Thank you both for your feedback! I have generated a (hopefully better) test case in `GlobalISel/legalize-ashr.mir`.

arsenm commented on llvm/test/CodeGen/AMDGPU/widen-vector-shift.ll:
```diff
@@ -0,0 +1,24 @@
+; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx90a -O0 -print-after=legalizer %s -o /dev/null 2>&1 | FileCheck %s
+
+; CHECK-LABEL: widen_ashr_i4:
+define amdgpu_kernel void @widen_ashr_i4(
+    ptr addrspace(1) %res, i4 %a, i4 %b) {
+; CHECK: G_ASHR %{{[0-9]+}}:_, %{{[0-9]+}}:_(s16)
+entry:
+  %res.val = ashr i4 %a, %b
+  store i4 %res.val, ptr addrspace(1) %res
+  ret void
```
Avoid the kernel and store if it's not relevant to the test

zGoldthorpe committed:
Integrated and autogenerated test.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-ashr.mir:
```diff
@@ -969,6 +969,176 @@ body: |
 
 ...
 
+---
```
I didn't mean switch this to a MIR test. This should still be and end to end IR test

shiltian replied:
It is testing the legalizer so it doesn't have to?

arsenm replied:
At this point we are over-using MIR tests which are bigger, less maintainable, and test less. They are useful for testing edge cases that are difficult to reproduce, but simple stuff like this are better off just using IR. MIR legalizer tests were more useful at the start when we couldn't run as much code through end to end

shiltian replied:
The benefit of using MIR here is to be able to get auto generate check lines. Maybe we can teach the script to be able to generate check lines for llc test with IR input and MIR output.

arsenm replied:
You can generate ISA checks just fine, and update_llc_test_checks works much better than update_mir_test_checks

zGoldthorpe committed:
Rewrote tests in LLVM IR.

Also added tests for <4 x i2> vectors, which are still within the scope
of the patch, particularly because an arithmetic right shift is an
identity transformation.
